docs/design/PT-166642114-miner-activated-consensus.md

```diff
@@ -0,0 +1,146 @@
+with a storage access that enables non-transactional readers to see the outcome of the write
+immediately after the synchronous write call finishes from the point of view of the worker.
+
+Callers requiring the signalling outcome for a certain block
+shall enquiry the chain storage and, if no outcome yet,
+ensure that there is a worker alive for that
+then yield (e.g. for sync worker process, for locally mined key blocks) or terminate (e.g. for block from user API or gossip).
+
+##### Blocks not necessarily connected to genesis
+
+The current design of syncing receives blocks out of order.
+Therefore the validation of the consensus protocol version in each block at that stage needs to be relaxed.
+Any deferred validation must be performed on each single block when meant be inserted in the chain storage,
+as at that stage such block is meant to be connected to genesis.
+
+##### Blocks persisted in the database with a now-invalid consensus protocol
+
+**TODO Handling of blocks read from database after user configuration makes consensus upgrade unconditional.**
+
+#### Transactions
+
+##### Mempool
+
+###### Insertion
+
+When configured for miner signalling, then node relaxes checks on transactions
+so to accept both transactions valid under the current consensus protocol or under the proposed new consensus protocol.
+
+###### Deletion
+
+When configured for miner signalling,
+after - on the best valid chain - the proposed fork height is reached and the consensus protocol version is decided,
+the node will not re-check transactions for the now-invalid protocol.
+
+After having configured the node for miner signalling,
+if afterwards the user configures a fixed consensus protocol at a certain height and restarts the node
+the node will not re-check transactions for the now-invalid protocol.
+
+##### State channel client finite state machine
+
+**TODO (The state channel client finite state machine handles transactions assuming consensus protocol version.)**
+
+**TODO Handling of transactions read from database after user configuration makes consensus upgrade unconditional.**
+
+##### User API
+
+**TODO The user API returns transactions to the user. Mark those APIs as deprecated and remove those? Let user specify consensus protocol?**
+
+##### Other reads
+
+**TODO Handling of transactions read from database after user configuration makes consensus upgrade unconditional.**
+
+### Data structures
+
+#### User configuration
+
+User configuration for accommodating the [defined parameters][fspecs].
+**TODO Detail format and constraints. (See [comment Pivotal Tracker ticket](https://www.pivotaltracker.com/story/show/166642114/comments/205266011) and following comments.)**
+
+#### Auxiliary table for asynchronous computation of outcome of signalling
+
+Table whose name is computed based on the set of [user configured parameters][fspecs] - including current consensus protocol version C1, and excluding the height H1 at which it was first effective.
```

Comment:
The table might not be necessary as we could recompute the result and keep it in a worker process.

Reply:
Or an in-memory table.

Reply:
Anyway the lookup table/dict for reaching the process is required.

So it looks like that, once worker finished computing outcome, outcome can be kept in some in-memory state (either in conductor state as dict, or in ETS conditionally created by conductor init, or in *overcomplicated* non-persisted Mnesia table).

Reply:
Conductor's state is presumably the simplest option.

Reply:
Unresolved so to keep track of this (I have not addressed yet).

Reply:
Addressed in 5f917bbc8